android/app/src/main/cpp/saf_wrapper.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2020 Taner Sener
+ *
+ * This file is part of MobileFFmpeg.
+ *
+ * MobileFFmpeg is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * MobileFFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with MobileFFmpeg.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <sys/stat.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "config.h"
+#include "libavformat/avformat.h"
+#include "libavutil/avstring.h"
+
+#include "mobileffmpeg.h"
+#include "saf_wrapper.h"
+
+// in these wrappers, we call the original functions, so we remove the shadow defines
+#undef avio_closep
+#undef avformat_close_input
+#undef avio_open
+#undef avio_open2
+#undef avformat_open_input
+
+static int fd_read_packet(void* opaque, uint8_t* buf, int buf_size) {
+    int fd = (int)opaque;
+    return read(fd, buf, buf_size);
+}
+
+static int fd_write_packet(void* opaque, uint8_t* buf, int buf_size) {
+    int fd = (int)opaque;
+    return write(fd, buf, buf_size);
+}
+
+static int64_t fd_seek(void *opaque, int64_t offset, int whence) {
+    int fd = (int)opaque;
+
+    if (fd < 0) {
+        return AVERROR(EINVAL);
+    }
+    LOGD("fd_seek fd=%d %lld %d\n", fd, (long long)offset, whence);
+
+    int64_t ret;
+    if (whence == AVSEEK_SIZE) {
+        struct stat st;
+        ret = fstat(fd, &st);
+        return ret < 0 ? AVERROR(errno) : (S_ISFIFO(st.st_mode) ? 0 : st.st_size);
+    }
+
+    ret = lseek(fd, offset, whence);
+
+    return ret < 0 ? AVERROR(errno) : ret;
+}
+
+__thread char link_name_buf[1024];
+static const char *get_link_name(int fd) {
```

Comment:
Do we really need this method?

Reply:
It's important for debugging, but I will hide it if you prefer

Reply:
I have received permission denied errors when accessing `/proc/self/fd/` on some devices. And I suspect `readlink` method may fail on those. If `readlink` fails, `nbytes` will have a negative value and `link_name_buf[nbytes]` may cause access violation errors. 

I just didn't want to take that risk just for debugging. If it is important you can put it back and try to handle the error case for `readlink`.

Reply:
As cursory google search shows that there actually are reports of this `readlink()` failing on some older devices. I agree that keeping this log line is not worth having to deal with all possible exceptions in this case.